lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -0,0 +1,487 @@
+        success = toLangTagResultLength && U_SUCCESS(error);
+        if (!success)
+        {
+            if (error == UErrorCode::U_ILLEGAL_ARGUMENT_ERROR)
+            {
+                AssertMsg(false, "uloc_toLanguageTag: error U_ILLEGAL_ARGUMENT_ERROR");
+            }
+            else
+            {
+                AssertMsg(false, "uloc_toLanguageTag: unexpected error (besides U_ILLEGAL_ARGUMENT_ERROR)");
+            }
+
+            goto LReturn;
+        }
+
+    LReturn:
+        return success;
+    }
+
+    HRESULT NormalizeLanguageTag(_In_z_ const char16 *languageTag, _In_ const charcount_t cch,
+        _Out_ char16 *normalized, _Out_ size_t *normalizedLength)
+    {
+        bool success = false;
+        UErrorCode error = UErrorCode::U_ZERO_ERROR;
+        char icuLocaleId[ULOC_FULLNAME_CAPACITY] = { 0 };
+        char icuLangTag[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t parsedLength = 0;
+        int32_t forLangTagResultLength = 0;
+        int32_t toLangTagResultLength = 0;
+
+        // Allocate memory for the UTF8 output buffer. Need 3 bytes for each (code point + null) to satisfy SAL.
+        const size_t inputLangTagUtf8SizeAllocated = AllocSizeMath::Mul(AllocSizeMath::Add(cch, 1), 3);
+        unsigned char *inputLangTagUtf8 = new unsigned char[inputLangTagUtf8SizeAllocated];
+        StringBufferAutoPtr<unsigned char> guard(inputLangTagUtf8);
+
+        if (!inputLangTagUtf8)
+        {
+            AssertOrFailFastMsg(false, "OOM: failed to allocate inputLangTagUtf8.");
+            goto LReturn;
+        }
+
+        // explicitly zero the allocated array
+        for (size_t index = 0; index < inputLangTagUtf8SizeAllocated; ++index)
+        {
+            inputLangTagUtf8[index] = 0;
+        }
+
+        utf8::EncodeIntoAndNullTerminate(inputLangTagUtf8, languageTag, cch);
+
+        // Convert input language tag to a locale ID for use in uloc_toLanguageTag API.
+        forLangTagResultLength = uloc_forLanguageTag(reinterpret_cast<const char *>(inputLangTagUtf8), icuLocaleId, ULOC_FULLNAME_CAPACITY, &parsedLength, &error);
+        success = forLangTagResultLength && parsedLength && U_SUCCESS(error);
+        if (!success)
+        {
+            AssertMsg(false, "uloc_forLanguageTag failed");
+            goto LReturn;
+        }
+
+        // Try to convert icuLocaleId (locale ID version of input locale string) to BCP47 language tag, using strict checks
+        toLangTagResultLength = uloc_toLanguageTag(icuLocaleId, icuLangTag, ULOC_FULLNAME_CAPACITY, TRUE, &error);
+        success = toLangTagResultLength && U_SUCCESS(error);
+        if (!success)
+        {
+            if (error == UErrorCode::U_ILLEGAL_ARGUMENT_ERROR)
+            {
+                AssertMsg(false, "uloc_toLanguageTag: error U_ILLEGAL_ARGUMENT_ERROR");
+            }
+            else
+            {
+                AssertMsg(false, "uloc_toLanguageTag: unexpected error (besides U_ILLEGAL_ARGUMENT_ERROR)");
+            }
```

Comment:
Marking WontFix for now because I don't want to undertake any superficial refactorings at this point.

---
In reply to: [141681648](https://github.com/Microsoft/ChakraCore/pull/3809#discussion_r141681648) [](ancestors = 141681648)